File: vertx-lang-scala/src/main/scala/io/vertx/lang/scala/ScalaDocGenerator.scala
/*
 * Copyright (c) 2011-2016 The original author or authors
 * ------------------------------------------------------
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * and Apache License v2.0 which accompanies this distribution.
 *
 *     The Eclipse Public License is available at
 *     http://www.eclipse.org/legal/epl-v10.html
 *
 *     The Apache License v2.0 is available at
 *     http://www.opensource.org/licenses/apache2.0.php
 *
 * You may elect to redistribute this code under either of these licenses.
 */

package io.vertx.lang.scala

import io.vertx.codegen.`type`._
import io.vertx.codetrans.CodeTranslator
import io.vertx.docgen.Coordinate
import io.vertx.docgen.DocGenerator
import javax.annotation.processing.ProcessingEnvironment
import javax.lang.model.element._
import javax.lang.model.`type`.ExecutableType
import javax.lang.model.`type`.TypeMirror

import io.vertx.codetrans.lang.scala.ScalaLang

import scala.collection.JavaConverters._

/**
  * @author <a href="mailto:jochen.mader@codecentric.de">Jochen Mader</a>
  */
class ScalaDocGenerator extends DocGenerator {
  private var factory: TypeMirrorFactory = _
  private var translator: CodeTranslator = _
  private var env: ProcessingEnvironment = _

  def init(processingEnv: ProcessingEnvironment) {
    factory = new TypeMirrorFactory(processingEnv.getElementUtils, processingEnv.getTypeUtils)
    translator = new CodeTranslator(processingEnv)
    env = processingEnv
  }

  def getName: String = "scala"

  def renderSource(elt: ExecutableElement, source: String): String = {
    val lang: ScalaLang = new ScalaLang
    try
      translator.translate(elt, lang)

    catch {
      case e: Exception => {
        e.printStackTrace()
        System.out.println("Cannot generate " + elt.getEnclosingElement.getSimpleName + "#" + elt.getSimpleName + " : " + e.getMessage)
        "Code not translatable"
      }
    }
  }

  def resolveTypeLink(elt: TypeElement, coordinate: Coordinate): String = {
    var `type`: TypeInfo = null
    try
      `type` = factory.create(elt.asType)

    catch {
      case e: Exception => {
        System.out.println("Could not resolve doc link for type " + elt.getQualifiedName)
        return null
      }
    }
    if ((`type`.getKind eq ClassKind.ENUM) && `type`.asInstanceOf[EnumTypeInfo].isGen) {
      var baselink: String = null
      if (coordinate == null) baselink = "../"
      else baselink = "../../" + coordinate.getArtifactId + "/"
      return baselink + "enums.html#" + elt.getSimpleName.toString
    }
    if (`type`.getKind eq ClassKind.DATA_OBJECT) {
      var baselink: String = null
      if (coordinate == null) baselink = "../"
      else baselink = "../../" + coordinate.getArtifactId + "/"
      return baselink + "dataobjects.html#" + elt.getSimpleName.toString
    }
    if (`type`.getKind eq ClassKind.API) {
      val api: ApiTypeInfo = `type`.getRaw.asInstanceOf[ApiTypeInfo]
      return "../../scaladoc/" + api.translateName("scala").replace('.', '/') + ".html"
    }
    null
  }

  def resolveMethodLink(elt: ExecutableElement, coordinate: Coordinate): String = {
    val typeElt: TypeElement = elt.getEnclosingElement.asInstanceOf[TypeElement]
    var link: String = resolveTypeLink(typeElt, coordinate)
    if (link != null) if (link.contains("cheatsheet")) link = link + '#' + java.beans.Introspector.decapitalize(elt.getSimpleName.toString.substring(3))
    else {
      var anchor: String = '#' + elt.getSimpleName.toString + "("
      val `type`: TypeMirror = elt.asType
      val methodType: ExecutableType = env.getTypeUtils.erasure(`type`).asInstanceOf[ExecutableType]

      methodType.getParameterTypes.asScala.zipWithIndex.foreach{
        case (v, i) => {
          if(i > 0) anchor += ",%20"
          anchor += v.toString
        }
      }

      anchor += ')'
      link = link + anchor
    }
    link
  }

  def resolveLabel(elt: Element, defaultLabel: String): String = {
    if (elt.getKind eq ElementKind.METHOD) {
      val `type`: TypeInfo = factory.create(elt.getEnclosingElement.asType)
      if (`type`.getKind eq ClassKind.DATA_OBJECT) {
        var name: String = elt.getSimpleName.toString
        if (name.startsWith("set") && name.length > 3 && Character.isUpperCase(name.charAt(3))) name = java.beans.Introspector.decapitalize(name.substring(3))
        return name
      }
    }
    defaultLabel
  }

  def resolveConstructorLink(elt: ExecutableElement, coordinate: Coordinate): String = {
    return "todo"
  }

  def resolveFieldLink(elt: VariableElement, coordinate: Coordinate): String = {
    return "todo"
  }
}

File: vertx-lang-scala-codegen/src/main/java/io/vertx/lang/scala/codegen/ScalaDocGenerator.java
package io.vertx.lang.scala.codegen;

import io.vertx.codegen.type.*;
import io.vertx.codetrans.CodeTranslator;
import io.vertx.codetrans.lang.scala.ScalaLang;
import io.vertx.docgen.Coordinate;
import io.vertx.docgen.DocGenerator;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.element.*;
import java.util.List;
import java.util.stream.IntStream;

/**
 * @author <a href="mailto:jochen@codepitbull.de">Jochen Mader</a>
 */
public class ScalaDocGenerator implements DocGenerator {
  private TypeMirrorFactory factory;
  private CodeTranslator translator;
  private ProcessingEnvironment env;

  @Override
  public void init(ProcessingEnvironment processingEnvironment) {
    factory = new TypeMirrorFactory(processingEnvironment.getElementUtils(), processingEnvironment.getTypeUtils());
    translator = new CodeTranslator(processingEnvironment);
    env = processingEnvironment;
  }

  @Override
  public String getName() {
    return "scala";
  }

  @Override
  public String renderSource(ExecutableElement elt, String source) {
    ScalaLang lang = new ScalaLang();
    try {
      return translator.translate(elt, lang);
    }
    catch (Exception e){
      e.printStackTrace();
      System.out.println("Cannot generate " + elt.getEnclosingElement().getSimpleName() + "#" + elt.getSimpleName() + " : " + e.getMessage());
      return "Code not translatable";
    }
  }

  @Override
  public String resolveTypeLink(TypeElement elt, Coordinate coordinate) {
    TypeInfo type = null;
    try {
      type = factory.create(elt.asType());
    }

    catch (Exception e) {
      System.out.println("Could not resolve doc link for type " + elt.getQualifiedName());
      return null;
    }
    if (type.getKind().equals(ClassKind.ENUM) && ((EnumTypeInfo)type).isGen()) {
      String baselink = null;
      if (coordinate == null) baselink = "../";
      else baselink = "../../" + coordinate.getArtifactId() + "/";
      return baselink + "enums.html#" + elt.getSimpleName().toString();
    }
    if (type.getDataObject() != null) {
      String baselink = null;
      if (coordinate == null) baselink = "../";
      else baselink = "../../" + coordinate.getArtifactId() + "/";
      return baselink + "dataobjects.html#" + elt.getSimpleName().toString();
    }
    if (type.getKind().equals(ClassKind.API)) {
      ApiTypeInfo api = (ApiTypeInfo)type.getRaw();
      return "../../scaladocs/" + api.translateName("scala").replace('.', '/') + ".html";
    }
    return null;
  }

  @Override
  public String resolveConstructorLink(ExecutableElement executableElement, Coordinate coordinate) {
    return "todo";
  }

  @Override
  public String resolveMethodLink(ExecutableElement elt, Coordinate coordinate) {
    TypeElement typeElt = (TypeElement)elt.getEnclosingElement();
    String link = resolveTypeLink(typeElt, coordinate);
    if (link != null) if (link.contains("cheatsheet")) link = link + '#' + java.beans.Introspector.decapitalize(elt.getSimpleName().toString().substring(3));
    else {

      StringBuilder anchor = new StringBuilder('#' + elt.getSimpleName().toString() + "(");
      TypeMirror type = elt.asType();
      ExecutableType methodType = (ExecutableType)env.getTypeUtils().erasure(type);

      List<? extends TypeMirror> parameterTypes = methodType.getParameterTypes();

      IntStream.range(0, parameterTypes.size()-1).forEach(index -> {
        if(index > 0) anchor.append(",%20");
        //drop preceding annotations as they mess up linking
        String[] splitted = parameterTypes.get(index).toString().split(" :: ");
        anchor.append(splitted[splitted.length - 1]);
      });

      anchor.append(')');
      link = link + anchor.toString();
    }
    return link;
  }

  @Override
  public String resolveFieldLink(VariableElement variableElement, Coordinate coordinate) {
    return "todo";
  }

  @Override
  public String resolveLabel(Element elt, String defaultLabel) {
    if (elt.getKind().equals(ElementKind.METHOD)) {
      TypeInfo type = factory.create(elt.getEnclosingElement().asType());
      if (type.getDataObject() != null) {
        String name = elt.getSimpleName().toString();
        if (name.startsWith("set") && name.length() > 3 && Character.isUpperCase(name.charAt(3))) name = java.beans.Introspector.decapitalize(name.substring(3));
        return name;
      }
    }
    return defaultLabel;
  }
}
